content/blog/analysis-with-RCrunch.Rmd

```diff
@@ -0,0 +1,300 @@
+
+## Setup
+First we need to load the Crunch package, sign in with our Crunch credentials, and then load the dataset that we would like to analyze. We are using data from [The Stack Overflow annual developer survey](https://insights.stackoverflow.com/survey/?utm_source=so-owned&utm_medium=blog&utm_campaign=dev-survey-2017&utm_content=blog-link&utm_term=data).
+```{r results='hide',message=F, warning=F}
+library(crunch)
+library(httpcache)
+login()
+so_data <- loadDataset("Stack Overflow Developer Survey (2017)")
+```
+
+```{r, include=FALSE}
+library(knitr) # for kables
+
+# reset category nas, values, and order to pre-cleaned state
+is.na(categories(so_data$HomeRemote)) <- c(rep(FALSE, 8), TRUE)
+# reset category values to pre-cleaned state
+indeces <- match(c(1:8, -1), ids(categories(so_data$HomeRemote)))
+categories(so_data$HomeRemote) <- categories(so_data$HomeRemote)[indeces]
+
+# undo variable name transformations
+var_names <- names(variables(so_data))
+var_names <- gsub(" - (numeric|text)", "", var_names)
+names(variables(so_data)) <- var_names
+
+# delete case variables
+try(with_consent(delete(so_data[['OScontributor']])))
+try(with_consent(delete(so_data[['HomeRemoteCleaned']])))
+so_data <- refresh(so_data)
+```
+
+Now our data is ready to be used with R in the `so_data` object. The first thing to note is that we haven't actually transferred all of the data in the dataset from Crunch servers to our local computer (which could take considerable time for large datasets). Instead, we've just transferred a small amount of metadata about the dataset that we can use to communicate with Crunch servers.
+
+## Dataset shape
+This dataset object is designed to use many of these same functions that dataframes and other R objects use. This helpful for getting features like dimensions of the whole data set.
+
+```{r}
+dim(so_data)
+```
+
+Using functions like this with datasets will generally be much faster than trying more roundabout ways of looking at data summaries. For example, if we wanted to know how many rows there are in our dataset, we could pull one of the columns, and then use `length` to count how many elements there are. 
+```{r}
+respondent_col <- as.vector(so_data$Respondent)
+length(respondent_col)
+```
+
+But we don't actually need to pull all of the data from a column just to know how many rows there are, we can use `nrow` with the dataset object, and the Crunch package is smart enough to use the metadata it already has to tell us how many rows there are.
+```{r}
+nrow(so_data)
+```
+
+The time savings are not trivial either, pulling the data:
+```{r}
+clearCache() # let's make this a fair fight
+system.time({
+    respondent_col <- as.vector(so_data$Respondent)
+    length(respondent_col)  
+})
+
+```
+
+Using the Crunch package's built-in tools, we see a speed up of approaching 5 times.
```

Comment:
Worth noting of course this dataset would fit in memory and be faster locally — of course we’re not comparing network speed to local speed, and these are trivially simple operations, etc; but wave hands in the direction of like
> Now consider that this is just the 2017 data, and Stack Overflow has been conducting this survey for 10 years. Thousands of columns, of hundreds of thousands or millions of rows, is not R’s strong suit, and we have been surprised at how quickly that limit can be reached with real-world data.